### PR TITLE
CNF-13709: Requeue and status handling in ClusterRequest controller

### PR DESCRIPTION
**Pull request by Missxiaoguo**
- Status conditions added
    - `ClusterRequestValidated`
             -  `Failed` if referenced ClusterTemplate is missing/invalid, merged clusterinstance input doesn't match schame, etc
    - `ClusterInstanceRendered`
             - `Failed` if ClusterInstance cannot be rendered
    - `ClusterResourcesCreated`
             - `Failed` if required resources(pull-secret, extramanifests configmaps) cannot be created due to inputError
    - `HardwareTemplateRendered`
             - `Failed` if hwtemplate cannot be rendered for NoodPool
    - `ClusterInstanceProcessed`
             - `NotApplied` if ClusterInstance could not be applied due to invalid/webhook issue
             - `Failed` if any condition (`ClusterInstanceValidated`, `RenderedTemplates`, `RenderedTemplatesValidated`, `RenderedTemplatesApplied`) from ClusterInstance CR is not completed
    - `HardwareProvisioned` (converted from status.HardwareProvisioningStatus)
    - `ClusterProvisioned`
            - Return status from the `Provisioned` condition in the ClusterInstance CR
            - `Unknown` if the Provisioned` in the ClusterInstance CR is missing
- TODO: add a status condition for configuration apply for day2 usecases
- Reconcilation tiggers for ClusterRequest
   - Creation, deletion and spec updates to the ClusterRequest
   - Status updates to the ClusterTemplate
   - status.conditions updates to the ClusterInstance
   - Status updates to the NoodPool
   - ztp-done label added to ManagedCluster
   - Internal errors could recover from a new reconcilation
- ClusterRequest labels are added to the created cluster namespace, ClusterInstance CR and NodePool CR for easier finalization and filtering of requests to be requeued
- Update to copy the list of configmaps specified in `extraManifestsRef` to cluster namespace if exist as  `extraManifestsRef` has changed to be a list in the siteconfig operator [AgentClusterInstall template](https://github.com/stolostron/siteconfig/blob/main/internal/templates/assisted-installer/template.go#L67-L70)

### CODE REVIEW

**Comment by Missxiaoguo** [comment]
Example of status output for success path
```
status:
  clusterInstanceRef:
    name: cnfdf20
    ztpStatus: ZTP Done
  conditions:
  - lastTransitionTime: "2024-08-13T04:39:44Z"
    message: The cluster request validation succeeded
    reason: Completed
    status: "True"
    type: ClusterRequestValidated
  - lastTransitionTime: "2024-08-13T04:39:44Z"
    message: Rendered ClusterInstance template successfully
    reason: Completed
    status: "True"
    type: ClusterInstanceRendered
  - lastTransitionTime: "2024-08-13T04:39:44Z"
    message: Cluster resources applied
    reason: Completed
    status: "True"
    type: ClusterResourcesCreated
  - lastTransitionTime: "2024-08-13T04:39:44Z"
    message: Rendered Hardware template successfully
    reason: Completed
    status: "True"
    type: HardwareTemplateRendered
  - lastTransitionTime: "2024-08-13T04:39:44Z"
    message: Processed ClusterInstance (cnfdf20) successfully
    reason: Completed
    status: "True"
    type: ClusterInstanceProcessed
  - lastTransitionTime: "2024-08-13T04:39:47Z"
    message: "Hareware provisioning completed"
    reason: Completed
    status: "True"
    type: HardwareProvisioned
  - lastTransitionTime: "2024-08-13T05:11:08Z"
    message: Provisioning completed
    reason: Completed
    status: "True"
    type: ClusterProvisioned
  nodePoolRef:
    name: cnfdf20
    namespace: dell-hwmgr
```

File: internal/controllers/clusterrequest_controller.go [hunk 1]
```diff
@@ -88,17 +95,25 @@ const (
 func (r *ClusterRequestReconciler) Reconcile(
 	ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
 	_ = log.FromContext(ctx)
+	result = doNotRequeue()
+
+	// Reconciliation loop can be triggered mulitiple times for the same resource
```
Comment (irinamihai) [inline internal/controllers/clusterrequest_controller.go:100]: Typo: "multiple"

File: internal/controllers/clusterrequest_controller.go [hunk 2]
```diff
@@ -119,50 +134,59 @@ func (r *ClusterRequestReconciler) Reconcile(
...
+			return requeueWithMediumInterval(), nil
+		}
+		return requeueWithError(err)
 	}
 
-	// ### HARDWARE TEMPLATE RENDERING ###
-	renderedNodePool, renderErr := t.renderHardwareTemplate(ctx, renderedClusterInstance)
-	if renderErr != nil {
-		return ctrl.Result{}, nil
+	// Render the hareware template for NodePool
```
Comment (irinamihai) [inline internal/controllers/clusterrequest_controller.go:177]: Typo: "hardware"

File: internal/controllers/clusterrequest_controller.go [hunk 3]
```diff
@@ -179,235 +203,303 @@ func (t *clusterRequestReconcilerTask) run(ctx context.Context) (nextReconcile c
...
-			slog.String("name", t.object.Name),
-		)
-		return
+		return fmt.Errorf("failed to get the ClusterTemplate for ClusterRequest %s: %w ", t.object.Name, err)
 	}
-	if createErr != nil {
-		return ctrl.Result{}, nil
+	ctValidatedCondition := meta.FindStatusCondition(clusterTemplate.Status.Conditions, string(utils.CTconditionTypes.Validated))
+	if ctValidatedCondition == nil || ctValidatedCondition.Status == metav1.ConditionFalse {
+		return utils.NewInputError("the clustertemplate validation is not passing")
```
Comment (irinamihai) [inline internal/controllers/clusterrequest_controller.go:264]: Maybe rephrase to `the clustertemplate validation has failed`

**Comment by Missxiaoguo** [comment]
Rebased. Testing with the latest code

**Comment by Missxiaoguo** [comment]
/unhold
Complete testing with latest code
